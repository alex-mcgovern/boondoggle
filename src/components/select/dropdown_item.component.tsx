import clsx from "clsx";
import React, { forwardRef } from "react";

import { Box } from "../box";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../..";
import type { DropdownItemShape } from "./select.types";
import type { Ref } from "react";

export interface DropdownItemProps {
  isHighlighted: boolean;
  isMultipleSelectionEnabled?: boolean;
  size?: SharedUiScale;
  isMulti?: boolean;
  item: DropdownItemShape;
  isDropdownItemSelected?: boolean;
}

export const DropdownItem = forwardRef(
  (
    {
      item,
      size,
      isHighlighted,
      isMulti,
      isDropdownItemSelected,
      ...rest
    }: DropdownItemProps,
    ref
  ) => {
    return (
      <Box
        as="button"
        id={item.label}
        className={clsx(
          styles.getDropdownItemStyles({
            size,
          }),
          {
            [styles.isHighlighted]: isHighlighted,
            [styles.isSelected]: isDropdownItemSelected,
          }
        )}
        {...rest}
        ref={ref as Ref<HTMLButtonElement>}
      >
        <Box flexShrink="0">{item.label}</Box>
        {isMulti && (
          <Box
            as="input"
            type="checkbox"
            readOnly
            tabIndex={-1}
            checked={isDropdownItemSelected}
            marginLeft="auto"
          />
        )}
      </Box>
    );
  }
);
