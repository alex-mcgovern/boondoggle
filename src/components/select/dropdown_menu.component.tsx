import { extractAtomsFromProps } from "@dessert-box/core";
import clsx from "clsx";
import React, { forwardRef } from "react";

import { getSprinkles } from "../../styles/utils/get_sprinkles.css";
import { Box } from "../box_component";
import { DropdownItem } from "./dropdown_item.component";
import * as styles from "./select.styles.css";

import type { SharedUiScale } from "../../styles/common/globalVariantsUiScale.css";
import type { SprinklesArgs } from "../../styles/utils/get_sprinkles.css";
import type { DropdownItemShape } from "./select.types";
import type {
  UseComboboxPropGetters,
  UseMultipleSelectionActions,
  UseMultipleSelectionGetSelectedItemPropsOptions,
} from "downshift";
import type { LegacyRef } from "react";

/**
 * ------------------------------------------------------------------------------
 * Util function to get props for individual dropdown items
 * May conditionally call `getItemProps` or `getSelectedItemProps` depending on
 * whether `isMulti` is true, and whether the item is selected or not
 * ------------------------------------------------------------------------------
 */

interface GetDropdownItemPropsArgs
  extends Pick<
    DropdownMenuProps,
    "getItemProps" | "getSelectedItemProps" | "removeSelectedItem"
  > {
  item: DropdownItemShape;
  isItemSelected?: boolean;
  [key: string]: unknown;
}

export const getDropdownItemProps = ({
  isItemSelected,
  getItemProps,
  getSelectedItemProps,
  removeSelectedItem,
  item,
  isMulti,
  ...rest
}: GetDropdownItemPropsArgs) => {
  if (isMulti && getSelectedItemProps) {
    return isItemSelected
      ? {
          ...getItemProps({
            item,
            ...rest,
          }),
          ...getSelectedItemProps({
            selectedItem: item,
            onClick: (e) => {
              e.stopPropagation();
              if (removeSelectedItem) {
                removeSelectedItem(item);
              }
            },
          }),
        }
      : getItemProps({ item, ...rest });
  }

  return getItemProps({ item });
};

/**
 * ------------------------------------------------------------------------------
 * Renders a dropdown menu for use with `SelectSingle` or `SelectMultiple`
 * ------------------------------------------------------------------------------
 */

export interface DropdownMenuProps extends SprinklesArgs {
  /** Required props  */
  getIsItemSelected: (item: DropdownItemShape) => boolean;
  getItemProps: UseComboboxPropGetters<DropdownItemShape>["getItemProps"];
  getMenuProps: UseComboboxPropGetters<DropdownItemShape>["getMenuProps"];
  isOpen: boolean;
  items: Array<DropdownItemShape>;

  /** Optional props  */
  getSelectedItemProps?: (
    options: UseMultipleSelectionGetSelectedItemPropsOptions<DropdownItemShape>
  ) => { [key: string]: unknown };
  isMulti?: boolean;
  highlightedIndex?: number | undefined;
  removeSelectedItem?: UseMultipleSelectionActions<DropdownItemShape>["removeSelectedItem"];
  size?: SharedUiScale;
}

export const DropdownMenu = forwardRef(
  (
    {
      getIsItemSelected,
      getItemProps,
      getMenuProps,
      getSelectedItemProps,
      highlightedIndex,
      removeSelectedItem,
      isMulti,
      isOpen,
      items,
      size = "md",
      ...rest
    }: DropdownMenuProps,
    ref
  ) => {
    const dropdownWrapperStyles = clsx(styles.dropdownListWrapper, {
      [styles.dropdownWrapperClosed]: !isOpen,
    });

    const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

    /**
     * Note: `DropdownMenu` *must* not be in a conditional render, or
     * downshift's `getMenuProps` will be unable to apply a ref and throw an error
     */
    return (
      <Box
        className={dropdownWrapperStyles}
        {...atomProps}
        {...getMenuProps?.({
          ...otherProps,
          ref: ref as LegacyRef<HTMLElement>,
        })}
      >
        {Array.isArray(items) &&
          items.map((item, index) => {
            if (!item.label) {
              return null;
            }

            const isItemSelected = getIsItemSelected?.(item);

            return (
              <DropdownItem
                item={item}
                isMulti={isMulti}
                isHighlighted={highlightedIndex === index}
                key={item.label}
                isDropdownItemSelected={isItemSelected}
                size={size}
                {...getDropdownItemProps({
                  isItemSelected,
                  index,
                  getItemProps,
                  getSelectedItemProps,
                  item,
                  isMulti,
                  removeSelectedItem,
                })}
              />
            );
          })}
      </Box>
    );
  }
);
