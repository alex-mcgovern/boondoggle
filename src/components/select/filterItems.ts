import type { SelectItemShape } from "./types";

type GetFilteredDropdownItemsArgs = {
  /** Value of controlled combobox input */
  inputValue?: string;
  /** An array of dropdown items */
  items: Array<SelectItemShape>;
};

/**
 * Removes items not matching `inputValue` from `items`
 */
export function filterItems({
  inputValue,
  items,
}: GetFilteredDropdownItemsArgs) {
  if (!inputValue) {
    return items;
  }

  return items.filter((item) => {
    return (
      item.label.toLowerCase().includes(inputValue.toLowerCase()) ||
      item.value.toLowerCase().includes(inputValue.toLowerCase())
    );
  });
}
