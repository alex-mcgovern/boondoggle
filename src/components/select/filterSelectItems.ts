import { arrayHasLength } from "../../lib/array_has_length";
import { isTruthy } from "../../lib/is_truthy";
import { isFlatSelectItems } from "./isFlatSelectItems";
import { isGroupedSelectItems } from "./isGroupedSelectItems";

import type { FlatSelectItems, GroupedSelectItems } from "./types";

/**
 * Removes items not matching `inputValue` from `items`
 */
export function filterSelectItems<
    TValue extends string = string,
    TItemData extends Record<string, unknown> = Record<string, unknown>
>({
    inputValue,
    items,
}: {
    /**
     * Value of controlled combobox input
     */
    inputValue?: string;

    /**
     * An array of dropdown items
     */
    items:
        | FlatSelectItems<TValue, TItemData>
        | GroupedSelectItems<TValue, TItemData>
        | undefined;
}):
    | FlatSelectItems<TValue, TItemData>
    | GroupedSelectItems<TValue, TItemData>
    | undefined {
    if (!inputValue) {
        return items;
    }

    if (isFlatSelectItems(items)) {
        items.filter((item) => {
            return (
                item.label.toLowerCase().includes(inputValue.toLowerCase()) ||
                item.value.toLowerCase().includes(inputValue.toLowerCase()) ||
                item.tags?.some((tag) => {
                    return tag.toLowerCase().includes(inputValue.toLowerCase());
                })
            );
        });
    }

    if (isGroupedSelectItems(items)) {
        return items
            .map((group) => {
                const filteredWithinGroup = group.items.filter((item) => {
                    return (
                        item.label
                            .toLowerCase()
                            .includes(inputValue.toLowerCase()) ||
                        item.value
                            .toLowerCase()
                            .includes(inputValue.toLowerCase()) ||
                        item.tags?.some((tag) => {
                            return tag
                                .toLowerCase()
                                .includes(inputValue.toLowerCase());
                        })
                    );
                });

                if (!arrayHasLength(filteredWithinGroup)) {
                    return null;
                }

                return {
                    ...group,
                    items: filteredWithinGroup,
                };
            })
            .filter(isTruthy);
    }

    return items;
}
