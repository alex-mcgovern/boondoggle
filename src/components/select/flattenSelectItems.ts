import { isFlatSelectItems } from "./isFlatSelectItems";

import type {
    FlatSelectItems,
    GroupedSelectItems,
    SelectItemShape,
} from "./types";

export function flattenSelectItems<
    TValue extends string = string,
    TItemData extends Record<string, unknown> = Record<string, unknown>
>(
    items:
        | FlatSelectItems<TValue, TItemData>
        | GroupedSelectItems<TValue, TItemData>
        | undefined
) {
    if (!items) {
        return [];
    }

    if (isFlatSelectItems(items)) {
        return items;
    }

    return items.reduce(
        (prev: Array<SelectItemShape<TValue, TItemData>>, acc) => {
            return [...prev, ...acc.items];
        },
        []
    );
}
