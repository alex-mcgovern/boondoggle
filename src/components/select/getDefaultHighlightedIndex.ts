import type { SelectItemShape } from "./types";

type GetDefaultHighlightedIndexArgs = {
  /** The initial highlighted item */
  initialHighlightedItem?: SelectItemShape;
  /** An array of dropdown items */
  items: Array<SelectItemShape>;
  /** The selected item */
  selectedItem?: SelectItemShape | null;
};

/**
 * Util to get the default highlighted index for a dropdown
 */
export const getDefaultHighlightedIndex = ({
  initialHighlightedItem,
  items,
  selectedItem,
}: GetDefaultHighlightedIndexArgs) => {
  if (selectedItem) {
    return items.findIndex((item) => {
      return item.label === selectedItem?.label;
    });
  }

  return items.findIndex((item) => {
    return item.label === initialHighlightedItem?.label;
  });
};
