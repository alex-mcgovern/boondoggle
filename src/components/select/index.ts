export { SelectSingle } from "./select_single.comp";
export type { SelectSingleProps } from "./select_single.comp";

export { SelectButton } from "./select_button.comp";
export type { SelectButtonProps } from "./select_button.comp";

export { SelectMulti } from "./select_multi.comp";
export type { SelectMultiProps } from "./select_multi.comp";

export type { DropdownItemShape } from "./select.types";
