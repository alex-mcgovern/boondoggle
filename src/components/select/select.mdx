import { ArgTypes, Canvas, Meta, Story } from "@storybook/addon-docs";

import * as stories from "./select.stories.tsx";

<Meta of={stories} />

# Select

- Wrapper around Downshift to provide a select component with a few options for
  styling and behavior

## Select Single

<Canvas name="default" of={stories.SelectSingleDefault} />

## Select multiple

<Canvas name="default" of={stories.SelectMultipleDefault} />

## Filterable

[ToDo: Add filterable example]
{/* <Canvas name="filterable" of={stories.Filterable} /> */}
