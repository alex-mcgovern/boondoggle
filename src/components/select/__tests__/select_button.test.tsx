/** @jest-environment jsdom */
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { SelectButton } from "..";
import { LOREM } from "../../../../mocks/LOREM.mock";
import { elementFontSize } from "../../../styles/common/element_size.css";
import { Icon } from "../../icon";
import { mockSelectItems } from "../__mocks__/select.mock";

import type { SelectButtonProps } from "..";

const ON_CHANGE = jest.fn();
const ON_IS_OPEN_CHANGE = jest.fn();

const ITEMS = mockSelectItems({});
const FIRST_ITEM = ITEMS[0];
const FIRST_ITEM_LABEL = FIRST_ITEM?.label || "";
const SECOND_ITEM = ITEMS[1];
const SECOND_ITEM_LABEL = SECOND_ITEM?.label || "";

const PROPS: SelectButtonProps = {
  buttonText: LOREM.select,
  id: LOREM.id(),
  items: ITEMS,
  name: LOREM.textXxs,
  onChange: ON_CHANGE,
};

const renderComponent = async (props: SelectButtonProps) => {
  /**
   * Popper.js calls its update() method asynchronously after the
   * initial render, so we must wrap the render() call in waitFor()
   * to prevent test execution from being out of order.
   */
  return waitFor(async () => {
    return render(<SelectButton {...props} />);
  });
};

describe("<SelectButton />", () => {
  describe("Basic smoke tests", () => {
    it("should render without throwing", async () => {
      const { getByRole } = await renderComponent(PROPS);

      expect(getByRole("combobox")).not.toBeNull();
    });

    it("should render buttonText", async () => {
      const { getByRole } = await renderComponent(PROPS);

      const combobox = getByRole("combobox");
      expect(combobox as HTMLInputElement).toHaveTextContent(
        PROPS.buttonText as string
      );
    });

    it("should match snapshot", async () => {
      const { getByRole } = await renderComponent(PROPS);

      expect(getByRole("combobox")).toMatchSnapshot();
    });
  });

  describe("Disabled state", () => {
    it("should not show dropdown menu when user clicks", async () => {
      const { getByRole } = await renderComponent({ ...PROPS, disabled: true });

      const combobox = getByRole("combobox");
      await userEvent.click(combobox);

      const menu = getByRole("listbox", { hidden: true });
      expect(menu).not.toBeVisible();
    });

    it("should not show dropdown menu when user attempts keyboard navigation", async () => {
      const { getByRole } = await renderComponent({ ...PROPS, disabled: true });

      await userEvent.tab();
      await userEvent.keyboard("{arrowdown}");

      const menu = getByRole("listbox", { hidden: true });
      expect(menu).not.toBeVisible();
    });
  });

  describe("Dropdown menu", () => {
    it("should not be visible on first mount", async () => {
      const { getByRole } = await renderComponent(PROPS);

      const menu = getByRole("listbox", { hidden: true });
      expect(menu).not.toBeVisible();
    });

    it("should be visible after user clicks on select", async () => {
      const { getByRole } = await renderComponent(PROPS);

      const combobox = getByRole("combobox");
      await userEvent.click(combobox);

      const menu = getByRole("listbox");
      expect(menu).toBeVisible();
    });

    it("should be visible after user opens select with keyboard", async () => {
      const { getByRole } = await renderComponent(PROPS);

      await userEvent.tab();
      await userEvent.keyboard("{arrowdown}");

      const menu = getByRole("listbox");
      expect(menu).toBeVisible();
    });
  });

  describe("On `isOpen` change", () => {
    it("should call onIsOpenChange when user opens select by clicking", async () => {
      const { getByRole } = await renderComponent({
        ...PROPS,
        onIsOpenChange: ON_IS_OPEN_CHANGE,
      });

      const combobox = getByRole("combobox");
      await userEvent.click(combobox);

      expect(ON_IS_OPEN_CHANGE).toHaveBeenCalledWith(
        expect.objectContaining({
          isOpen: true,
        })
      );
    });

    it("should call onIsOpenChange when user opens select with keyboard", async () => {
      renderComponent({
        ...PROPS,
        onIsOpenChange: ON_IS_OPEN_CHANGE,
      });

      await userEvent.tab();
      await userEvent.keyboard("{arrowdown}");

      expect(ON_IS_OPEN_CHANGE).toHaveBeenCalledWith(
        expect.objectContaining({
          isOpen: true,
        })
      );
    });
  });

  describe("Selecting with mouse", () => {
    it("should call `onChange` with first clicked item", async () => {
      const { getByRole, getByText } = await renderComponent(PROPS);

      const combobox = getByRole("combobox");
      await userEvent.click(combobox);

      const firstItem = getByText(FIRST_ITEM_LABEL);
      await userEvent.click(firstItem);

      expect(ON_CHANGE).toHaveBeenCalledWith(
        expect.objectContaining({
          selectedItem: FIRST_ITEM,
        })
      );
    });

    it("should call `onChange` with second clicked item", async () => {
      const { getByRole, getByText } = await renderComponent(PROPS);
      const combobox = getByRole("combobox");
      await userEvent.click(combobox);

      const firstItem = getByText(FIRST_ITEM_LABEL);
      await userEvent.click(firstItem);

      const secondItem = getByText(SECOND_ITEM_LABEL);
      await userEvent.click(secondItem);

      expect(ON_CHANGE).toHaveBeenLastCalledWith(
        expect.objectContaining({
          selectedItem: SECOND_ITEM,
        })
      );
    });
  });

  describe("Selecting with keyboard", () => {
    it("should call `onChange` with first item selected with keyboard", async () => {
      renderComponent(PROPS);

      await userEvent.tab();
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{enter}");

      expect(ON_CHANGE).toHaveBeenCalledWith(
        expect.objectContaining({
          selectedItem: FIRST_ITEM,
        })
      );
    });

    it("should call `onChange` with second item selected with keyboard", async () => {
      renderComponent(PROPS);

      await userEvent.tab();
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{enter}");

      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{arrowdown}");
      await userEvent.keyboard("{enter}");

      expect(ON_CHANGE).toHaveBeenLastCalledWith(
        expect.objectContaining({
          selectedItem: SECOND_ITEM,
        })
      );
    });
  });

  describe("Size", () => {
    it("should have the `md` class name by default", async () => {
      const { getByRole } = await renderComponent(PROPS);

      expect(getByRole("combobox")).toHaveClass(elementFontSize.md);
    });

    it("should have the `sm` class name when size = sm", async () => {
      const { getByRole } = await renderComponent({ ...PROPS, size: "sm" });

      expect(getByRole("combobox")).toHaveClass(elementFontSize.sm);
    });

    it("should have the `md` class name when size = md", async () => {
      const { getByRole } = await renderComponent({ ...PROPS, size: "md" });

      expect(getByRole("combobox")).toHaveClass(elementFontSize.md);
    });

    it("should have the `lg` class name when size = lg", async () => {
      const { getByRole } = await renderComponent({ ...PROPS, size: "lg" });

      expect(getByRole("combobox")).toHaveClass(elementFontSize.lg);
    });
  });

  describe("Slot props", () => {
    it("should render node passed to `slotLeft`", async () => {
      const { getByTestId } = await renderComponent({
        ...PROPS,
        slotLeft: <Icon data-testid="icon" icon={faSearch} />,
      });

      expect(getByTestId("icon")).not.toBeNull();
    });

    it("should render node passed to `slotRight`", async () => {
      const { getByTestId } = await renderComponent({
        ...PROPS,
        slotRight: <Icon data-testid="icon" icon={faSearch} />,
      });
      expect(getByTestId("icon")).not.toBeNull();
    });
  });
});
