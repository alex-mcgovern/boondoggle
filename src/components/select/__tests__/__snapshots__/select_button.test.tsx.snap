// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectButton /> Basic smoke tests should match snapshot 1`] = `
<button
  aria-activedescendant=""
  aria-controls="downshift-10-menu"
  aria-expanded="false"
  aria-haspopup="listbox"
  aria-labelledby="downshift-10-label"
  class="button.styles_getButtonStyles__3vuwsi8 a11y_a11yFocus__r8he5t2 a11y_a11yDisabled__r8he5t1 button.styles__3vuwsi6 get_sprinkles_alignItems_center__eiij7b1c get_sprinkles_borderRadius_md__eiij7ba get_sprinkles_display_flex_mobile__eiij7bdc get_sprinkles_gap_spacing1__eiij7b1r get_sprinkles_textDecoration_none__eiij7b9q button.styles_getButtonStyles_appearance_primary__3vuwsi9 button.styles_variantAppearance_primary__3vuwsi3 button.styles__3vuwsi0 get_sprinkles_paddingLeft_spacing2__eiij7b7i get_sprinkles_paddingRight_spacing2__eiij7b7s get_sprinkles_paddingTop_spacing1__eiij7b81 get_sprinkles_paddingBottom_spacing1__eiij7b77 get_sprinkles_justifyContent_center__eiij7b2k get_sprinkles_whiteSpace_nowrap__eiij7b9u get_sprinkles_fontWeight_semibold__eiij7b2v button.styles_getButtonStyles_size_md__3vuwsie element_size_elementSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_lg__eiij7b8g get_sprinkles_lineHeight_body_lg__eiij7b8r"
  id="lorem-ipsum"
  name="Lorem ipsum"
  role="combobox"
  tabindex="0"
  type="button"
>
  <div
    class="get_sprinkles_display_flex_mobile__eiij7bdc get_sprinkles_alignItems_start__eiij7b1b get_sprinkles_gap_spacing2__eiij7b1s get_sprinkles_color_inherit__eiij7b19 box.styles_box__1xdhb490"
  >
    Select a thing
    <div
      class="get_sprinkles_paddingTop_spacing1__eiij7b81 get_sprinkles_paddingBottom_spacing1__eiij7b77 get_sprinkles_color_inherit__eiij7b19 get_sprinkles_width_spacing3_mobile__eiij7bbu box.styles_box__1xdhb490"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-down icon.styles_icon__15uu6ts0 get_sprinkles_flexShrink_0__eiij7b1m"
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</button>
`;
